frame/contracts/src/migration.rs

```diff
@@ -24,59 +27,117 @@ use frame_support::{
 }
 
-trait MigrateSequence<T: Config> {
-	const VERSION_RANGE: Option<(u16, u16)>;
+/// A noop migration that can be used when there is no migration to be done for a given version.
+#[derive(frame_support::DefaultNoBound, Encode, Decode, MaxEncodedLen)]
+pub struct NoopMigration<const N: u16>;
+
+impl<const N: u16> Migrate for NoopMigration<N> {
+	const VERSION: u16 = N;
+	fn max_step_weight() -> Weight {
+		Weight::zero()
+	}
+	fn step(&mut self) -> (IsFinished, Weight) {
+		log::debug!(target: LOG_TARGET, "No migration for version {}", N);
+		(IsFinished::Yes, Weight::zero())
+	}
+}
 
+mod private {
+	pub trait Sealed {}
+	#[impl_trait_for_tuples::impl_for_tuples(10)]
+	#[tuple_types_custom_trait_bound(crate::Migrate)]
+	impl Sealed for Tuple {}
+}
+
+/// Defines a sequence of migrations.
+/// The sequence must be defined by a tuple of migrations, each of which must implement the
+/// `Migrate` trait. Migrations must be ordered by their versions with no gaps.
+pub trait MigrateSequence: private::Sealed {
+	/// Returns the range of versions that this migration can handle.
+	/// Migrations must be ordered by their versions with no gaps.
+	/// The following code will fail to compile:
+	///
+	/// The following code will fail to compile:
+	/// ```compile_fail
+	///     # use pallet_contracts::{NoopMigration, MigrateSequence};
+	/// 	let _ = <(NoopMigration<1>, NoopMigration<3>)>::VERSION_RANGE;
+	/// ```
+	/// The following code will compile:
+	/// ```
+	///     # use pallet_contracts::{NoopMigration, MigrateSequence};
+	/// 	let _ = <(NoopMigration<1>, NoopMigration<2>)>::VERSION_RANGE;
+	/// ```
+	const VERSION_RANGE: (u16, u16);
+
+	/// Returns the default cursor for the given version.
 	fn new(version: StorageVersion) -> Cursor;
 
+	/// Execute the migration step until the weight limit is reached.
+	/// Returns the new cursor or `None` if the migration is finished.
 	fn steps(version: StorageVersion, cursor: &[u8], weight_left: &mut Weight) -> Option<Cursor>;
 
+	/// Verify that each migration's cursor fits into the Cursor type.
 	fn integrity_test();
 
+	/// Returns whether migration from `in_storage` to `target` is supported.
+	/// A migration is supported if (in_storage + 1, target) is contained by `VERSION_RANGE`.
 	fn is_upgrade_supported(in_storage: StorageVersion, target: StorageVersion) -> bool {
 		if in_storage == target {
 			return true
 		}
 		if in_storage > target {
 			return false
 		}
-		let Some((low, high)) = Self::VERSION_RANGE else {
-			return false
-		};
+
+		let (low, high) = Self::VERSION_RANGE;
 		let Some(first_supported) = low.checked_sub(1) else {
 			return false
 		};
-		in_storage == first_supported && target == high
+		in_storage >= first_supported && target == high
 	}
 }
 
 /// Performs all necessary migrations based on `StorageVersion`.
-pub struct Migration<T: Config>(PhantomData<T>);
+pub struct Migration<T: Config, M: MigrateSequence = Migrations<T>>(PhantomData<(T, M)>);
```

Comment:
Made Migrations<T> configurable so we can write some tests on a Mocked migrations